### PR TITLE
[DataPipe] Unifying API - removing options to have fn_args and fn_kwargs from IterDataPipes

### PR DESCRIPTION
**Pull request by NivekT**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69561
* __->__ #69560

cc @VitalyFedyunin @ejguan @NivekT

Differential Revision: [D32952100](https://our.internmc.facebook.com/intern/diff/D32952100)

### CODE REVIEW

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32952100).

File: test/test_datapipe.py [hunk 1]
```diff
@@ -669,10 +669,10 @@ def _test_picklable(self):
         arr = range(10)
         picklable_datapipes: List[Tuple[Type[IterDataPipe], IterDataPipe, Tuple, Dict[str, Any]]] = [
             (dp.iter.Mapper, dp.iter.IterableWrapper(arr), (), {}),
-            (dp.iter.Mapper, dp.iter.IterableWrapper(arr), (_fake_fn, (0, ), {'test': True}), {}),
+            (dp.iter.Mapper, dp.iter.IterableWrapper(arr), (_fake_fn, (0, )), {}),
             (dp.iter.Collator, dp.iter.IterableWrapper(arr), (), {}),
-            (dp.iter.Collator, dp.iter.IterableWrapper(arr), (_fake_fn, (0, ), {'test': True}), {}),
-            (dp.iter.Filter, dp.iter.IterableWrapper(arr), (_fake_filter_fn, (0, ), {'test': True}), {}),
+            (dp.iter.Collator, dp.iter.IterableWrapper(arr), (_fake_fn, (0, )), {}),
+            (dp.iter.Filter, dp.iter.IterableWrapper(arr), (_fake_filter_fn, (0, )), {}),
```
Comment (ejguan) [inline test/test_datapipe.py:675]: Could we also add a picklable test for partial function?

File: test/test_datapipe.py [hunk 2]
```diff
@@ -1051,11 +1051,6 @@ def fn(item, dtype=torch.float, *, sum=False):
         for x, y in zip(map_dp, input_dp):
             self.assertEqual(x, torch.tensor(y, dtype=torch.float))
 
-        map_dp = input_dp.map(fn=fn, fn_args=(torch.int, ), fn_kwargs={'sum': True})
-        self.assertEqual(len(input_dp), len(map_dp))
-        for x, y in zip(map_dp, input_dp):
-            self.assertEqual(x, torch.tensor(y, dtype=torch.int).sum())
-
         from functools import partial
```
Comment (ejguan) [inline test/test_datapipe.py:1054]: Let's move the import to the top of this file.

File: test/test_datapipe.py [hunk 3]
```diff
@@ -1343,11 +1338,12 @@ def _filter_fn(data, val, clip=False):
                 return data >= val
             return True
 
-        filter_dp = input_ds.filter(filter_fn=_filter_fn, fn_args=(5, ))
+        from functools import partial
```
Comment (ejguan) [inline test/test_datapipe.py:1341]: Ditto

**Review by ejguan — APPROVED** [review]
LGTM with several nit comments.

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32952100).

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32952100).

**Comment by pmeier** [comment]
@NivekT @ejguan Please do not merge something like this without pinging me, because it breaks our CI and in turn will break `torchdata` CI.

**Comment by VitalyFedyunin** [comment]
@pmeier feel free to add ci rule to our github actions to control this.

**Comment by pmeier** [comment]
@VitalyFedyunin there already is a CI workflow for this [in `torchdata`](https://github.com/pytorch/data/blob/25530cf95878047f23e8f820d0ac5a00d539364f/.github/workflows/domain_ci.yml#L13-L58). Are you suggesting to add such a workflow also to PyTorch core? If yes, what would that entail? Build `torch` and `torchvision` from source and run the `torchvision` prototype datasets tests?

On a side note: why is there still development happening on this in PyTorch core? Can't we just move all development of this into `torchdata`?

**Comment by NivekT** [comment]
> @NivekT @ejguan Please do not merge something like this without pinging me, because it breaks our CI and in turn will break `torchdata` CI.

Sorry about that. The tests did not find any usage by `TorchVision`, I did find out that it breaks `TorchRec` so I worked with that team accordingly. They have some tests within the internal CI (not on Github). If it is burdensome to add more tests on PyTorch's Github CI, perhaps that can be another acceptable approach for testing?

> On a side note: why is there still development happening on this in PyTorch core? Can't we just move all development of this into `torchdata`?

There are DataPipes that will stay in PyTorch core permanently. These are some of them. We are changing their APIs slightly to unify them with the APIs of other libraries, such as `TorchArrow`. We should not expect a lot of changes but there may be some.

**Pull request closed without merging**